### PR TITLE
On macOS, register cmd+q on sclang interpreter main menu, solves #3824

### LINKED ISSUE #3824: Interpreter crashes when in focus and cmd-q is pressed (macOS)
Does this happen to anyone else?
Reproducer:
```
Window.new.front;
//now press Cmd-q (or Ctrl-q?)
//result:
//Interpreter has crashed or stopped forcefully. [Exit code: 11]
```
This will also happen without a gui window, as long as the interpreter (as opposed to the ide) is in focus.

Tested on recent develop, built against QT 5.9.1 on macOS. Some info from the console:
```
Exception Type:        EXC_BAD_ACCESS (SIGSEGV)
Exception Codes:       KERN_INVALID_ADDRESS at 0x0000000000000078
Exception Note:        EXC_CORPSE_NOTIFY

Termination Signal:    Segmentation fault: 11
Termination Reason:    Namespace SIGNAL, Code 0xb
Terminating Process:   exc handler [0]
```
```
Thread 2 Crashed:
0   sclang                        	0x000000010d034e68 boost::asio::detail::kqueue_reactor::run(long, boost::asio::detail::op_queue<boost::asio::detail::scheduler_operation>&) + 40 (conditionally_enabled_mutex.hpp:53)
1   sclang                        	0x000000010d034c59 boost::asio::detail::scheduler::do_run_one(boost::asio::detail::conditionally_enabled_mutex::scoped_lock&, boost::asio::detail::scheduler_thread_info&, boost::system::error_code const&) + 313 (scheduler.ipp:36)
2   sclang                        	0x000000010d034969 boost::asio::detail::scheduler::run(boost::system::error_code&) + 201 (scheduler.ipp:154)
3   sclang                        	0x000000010cfb4739 asioFunction() + 73 (error_code.hpp:494)
4   sclang                        	0x000000010cfa2298 void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void (*)()> >(void*) + 40 (memory:2600)
5   libsystem_pthread.dylib       	0x00007fffd534993b _pthread_body + 180
6   libsystem_pthread.dylib       	0x00007fffd5349887 _pthread_start + 286
7   libsystem_pthread.dylib       	0x00007fffd534908d thread_start + 13
```

### PR DESCRIPTION
**Pull request by geoffroymontel**
## Purpose and Motivation

Fixes #3824 

Like in PR #4500 the fix consists in adding an `Action` with `QAction::QuitRole` so that this action registers cmd+q on macOS.
I disable this menu action so that Quit menu item is greyed out in the interpreter window.

MainMenu is not affected, it still works, I checked.

## Types of changes

- Bug fix

## To-do list

- [ ] Code is tested
- [ ] All tests are passing
- [ ] Updated documentation
- [X] This PR is ready for review

### CODE REVIEW

**Comment by snappizz** [comment]
thanks geoffroy! cc @redFrik

**Comment by dyfer** [comment]
Thank you @geoffroymontel ! I won't be able to look at this for few days but I'd love to eventually - unless others get to it first :)

**Comment by dyfer** [comment]
I can confirm this prevents the crash when pressing `cmd+q`.

Not sure if it would be in the scope of this PR, but the SC GUI window still can't be closed with `cmd+w`. Would adding `cmd+w` to close the GUI window be a possible addition and in the scope of this PR?

File: QtCollider/QcApplication.cpp [hunk 1]
```diff
@@ -104,6 +104,29 @@ QcApplication::~QcApplication() {
...
+    QObject::connect(action, SIGNAL(triggered()), this, SLOT(onQuit()));
+    menu = new QMenu(tr("&File"));
+    menu->addAction(action);
+    _mainMenu->addMenu(menu);
+    action->setEnabled(false);
+#endif
+}
+
+void QcApplication::onQuit() {
+    printf("[QcApplication::onQuit] CMD+Q was caught on the interpreter. This is weird, it should not happen. \
```
Comment (dyfer) [inline QtCollider/QcApplication.cpp]: These are only wording suggestions, feel free to disregard:
`CMD+Q was caught by the Interpreter.`

File: QtCollider/QcApplication.cpp [hunk 2]
```diff
@@ -104,6 +104,29 @@ QcApplication::~QcApplication() {
...
+    Please fill an issue on https://github.com/supercollider/supercollider/issues");
```
Comment (dyfer) [inline QtCollider/QcApplication.cpp]: `Please file an issue at https://...`

**Comment by dyfer** [comment]
BTW should we write `cmd+q` or `cmd-q`, or all caps `CMD+Q`?

File: QtCollider/QcApplication.cpp (hunk 1 above)
Comment (geoffroymontel) [reply QtCollider/QcApplication.cpp]: That's better indeed, thanks !

File: QtCollider/QcApplication.cpp (hunk 2 above)
Comment (geoffroymontel) [reply QtCollider/QcApplication.cpp]: Thanks ! Solved.

**Comment by geoffroymontel** [comment]
Thanks for reviewing !!

> I can confirm this prevents the crash when pressing `cmd+q`.
Cool !

> Not sure if it would be in the scope of this PR, but the SC GUI window still can't be closed with `cmd+w` (#4309). Would adding `cmd+w` to close the GUI window be a possible addition and in the scope of this PR?

Mmm, it's quite different. Could you file a new issue with this feature request please ?

Best

**Comment by dyfer** [comment]
> Mmm, it's quite different. Could you file a new issue with this feature request please ?

#4309 is already filed, is this what you had in mind?

**Comment by geoffroymontel** [comment]
> #4309 is already filed, is this what you had in mind?
> 
Perfect ! Thanks for the pointer !

File: QtCollider/QcApplication.cpp [hunk 3]
```diff
@@ -104,6 +104,29 @@ QcApplication::~QcApplication() {
     _mutex.unlock();
 }
 
+void QcApplication::createMenu() {
+    _mainMenu = QSharedPointer<QMenuBar>(new QMenuBar(0));
```
Comment (brianlheim) [inline QtCollider/QcApplication.cpp]: The default argument for QMenuBar's constructor is nullptr, and QSharedPointer also has a more idiomatic (to modern c++) static factory function `create`, so this can be simplified to:

```cpp
_mainMenu = QSharedPointer<QMenuBar>::create();
```

File: QtCollider/QcApplication.cpp [hunk 4]
```diff
@@ -104,6 +104,29 @@ QcApplication::~QcApplication() {
...
+
+#ifdef Q_OS_MAC
+    // macOS registers cmd+q on menu bars by default. Here we register a handler for cmd+q that does nothing
+    // and we disable the Quit menu by default
+    QMenu* menu;
+    QAction* action;
+    action = new QAction(tr("Catch CMD+Q on MacOS X"));
```
Comment (brianlheim) [inline QtCollider/QcApplication.cpp]: We don't need to expose this or have a translation for it (`tr()` flags text as requiring a translation for the Qt meta-compiler), just replace it with an empty string.

File: QtCollider/QcApplication.cpp [hunk 5]
```diff
@@ -104,6 +104,29 @@ QcApplication::~QcApplication() {
...
+    menu = new QMenu(tr("&File"));
+    menu->addAction(action);
+    _mainMenu->addMenu(menu);
+    action->setEnabled(false);
+#endif
+}
+
+void QcApplication::onQuit() {
+    printf("[QcApplication::onQuit] CMD+Q was caught by the interpreter. This is weird, it should not happen. \
+    Please file an issue at https://github.com/supercollider/supercollider/issues");
```
Comment (brianlheim) [inline QtCollider/QcApplication.cpp]: this looks a bit odd, why have 5 spaces in a row? Probably what you mean is this:

```cpp
    printf("[QcApplication::onQuit] CMD+Q was caught by the interpreter. This is weird, it should not happen. "
    "Please file an issue at https://github.com/supercollider/supercollider/issues");
```

successive string literals (like `"abc" "def"`) are automatically concatenated by the compiler.

Comment (brianlheim) [reply QtCollider/QcApplication.cpp]: also, i believe you want `qWarning()` here, IIUC printf will not even work here. see for instance https://github.com/supercollider/supercollider/blob/develop/QtCollider/QtDownload.cpp#L77

File: QtCollider/QcApplication.cpp [hunk 6]
```diff
@@ -104,6 +104,29 @@ QcApplication::~QcApplication() {
...
+    // and we disable the Quit menu by default
+    QMenu* menu;
+    QAction* action;
+    action = new QAction(tr("Catch CMD+Q on MacOS X"));
+    action->setMenuRole(QAction::QuitRole);
+    QObject::connect(action, SIGNAL(triggered()), this, SLOT(onQuit()));
+    menu = new QMenu(tr("&File"));
+    menu->addAction(action);
+    _mainMenu->addMenu(menu);
+    action->setEnabled(false);
```
Comment (brianlheim) [inline QtCollider/QcApplication.cpp]: this block would be a lot clearer as:

```cpp
    auto* action = new QAction(...);
    action->setMenuRole
    action->setEnabled
    QObject::connect
    auto* menu = new QMenu
    menu->addAction
    _mainMenu->addMenu
```

in general - declare variables at the line they're needed and no earlier, use auto when the right-hand side type is clear, and group related calls together

File: QtCollider/QcApplication.cpp (hunk 3 above)
Comment (geoffroymontel) [reply QtCollider/QcApplication.cpp]: Thanks !

File: QtCollider/QcApplication.cpp (hunk 5 above)
Comment (geoffroymontel) [reply QtCollider/QcApplication.cpp]: Excellent ! The multiline string trick is good to know !

File: QtCollider/QcApplication.cpp (hunk 6 above)
Comment (geoffroymontel) [reply QtCollider/QcApplication.cpp]: Again, thanks, this is all very good to know !

**Comment by brianlheim** [comment]
Yes, it is. @geoffroymontel can you please fix that? Check the build log for the details.

**Comment by geoffroymontel** [comment]
@brianlheim sorry for that, it's fixed now !

**Comment by snappizz** [comment]
going through 3.10.4 changes for the changelog, i'm a bit confused by this one -- if cmd+q is supposed to quit the interpreter, why is it bad that the interpreter would quit with this shortcut? is this a regression from earlier behavior somehow?

**Comment by brianlheim** [comment]
see https://github.com/supercollider/supercollider/issues/3824#issuecomment-473657687

i'm not sure whether in the past cmd-q was supposed to quit the interpreter on macos. i think it's mainly that this can be confusing for beginners and doesn't offer much benefit for experienced users. someone else can probably answer this more fully though, i'm now also a bit confused myself.

**Comment by snappizz** [comment]
it also looks like the bug report in #3824 is a full-blown segmentation fault, not a graceful exit, but maybe someone could answer to what the behavior was before the crash (assuming that it was a regression).

**Comment by dyfer** [comment]
I is a regression. And yes, the issue was that this was a segfault instead of a graceful exit.

In 3.8, when the interpreter got a separate icon, cmd-q was not triggering shutdown (no action) when interpreter was in focus. This seems to be the desired behavior. At some point this changed to segfault, which is definitely a regression. This PR restores the behavior from 3.8.

For reference, in the cocoa days, and before the interpreter had a separate icon, there was no concept of shutting it off by itself. With either the editor or a gui window cmd-q would quit everything - editor, interpreter (and by extension the server). In practice, if one was in the middle of working on code with unsaved changes, any accidental cmd-q would prompt "save changes?" dialog, preventing accidental shutdown.
In short, cmd-q should not shut down the interpreter by itself. It should either not do anything (behavior with this PR), or possibly attempt shutting down the IDE (which would be closer to the cocoa SC behavior). But I really believe with this PR we have the desired behavior.

**Comment by snappizz** [comment]
thanks for the explanation.

there is a lot of "convention over configuration" keybindings happening in sclang gui, which i'm not a huge fan of. what if someone isn't using the IDE, and actually does want cmd+q to quit the interpreter?

**Comment by geoffroymontel** [comment]
I think supporting cmd+q to quit the interpreter causes more harm than benefits ! I don't see a practical use case for it. Removing it at least solves a few annoying bugs (like pressing cmd+q on a user Window during a performance would kill the interpreter).

**Comment by dyfer** [comment]
> there is a lot of "convention over configuration" keybindings happening in sclang gui, which i'm not a huge fan of. what if someone isn't using the IDE, and actually does want cmd+q to quit the interpreter?

I just checked - cmd+q is propagated as any other key combination to the views in the QT GUI. So user still has the option to implement quit if desired.
